### PR TITLE
HDDS-10054. Reduce DataNode token verification heap allocation cost.

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?
Remove unnecessary heap allocation associated with block token verification in DataNode.

Please describe your PR in detail:
Rewrite part of token verification code to use the underlying byte array directly. Avoid certain input/output stream objects that internally allocates sizable heap (4KB) each.

(I wanted to get rid of or optimize Token.decodeFromUrlString too, because BaseNCodec.ensureBufferSize allocates a 8192 byte buffer, which is overkill for this use case. This is another 1.x% overhead but I haven't find a way to remove it)

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10054

## How was this patch tested?

Ran with freon hsync workload and verified the change does remove heap allocation using profiler.

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/ContainerTokenIdentifier.java [hunk 1]
```diff
@@ -87,6 +87,16 @@ public void readFields(DataInput in) throws IOException {
...
   }
 
+  @Override
+  public void readFromByteArray(byte[] bytes) throws IOException {
+    ContainerTokenSecretProto proto =
+        ContainerTokenSecretProto.parseFrom(bytes);
+    setSecretKeyId(ProtobufUtils.fromProtobuf(proto.getSecretKeyId()));
+    setExpiry(Instant.ofEpochMilli(proto.getExpiryDate()));
+    setOwnerId(proto.getOwnerId());
+    this.containerID = ContainerID.getFromProtobuf(proto.getContainerId());
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/ContainerTokenIdentifier.java:97]: Please extract this part to a method and reuse in `readFields`:

https://github.com/apache/ozone/blob/41cb6d0ea95aeddc59e98005529204278e9be02f/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/ContainerTokenIdentifier.java#L84-L87

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java [hunk 2]
```diff
@@ -141,6 +141,18 @@ public void readFields(DataInput in) throws IOException {
...
+  @Override
+  public void readFromByteArray(byte[] bytes) throws IOException {
+    BlockTokenSecretProto token =
+        BlockTokenSecretProto.parseFrom(bytes);
+    setOwnerId(token.getOwnerId());
+    setExpiry(Instant.ofEpochMilli(token.getExpiryDate()));
+    setSecretKeyId(ProtobufUtils.fromProtobuf(token.getSecretKeyId()));
+    this.blockId = token.getBlockId();
+    this.modes = EnumSet.copyOf(token.getModesList());
+    this.maxLength = token.getMaxLength();
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java:153]: Same here, please share with:

https://github.com/apache/ozone/blob/41cb6d0ea95aeddc59e98005529204278e9be02f/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java#L136-L141

**Comment by adoroszlai** [comment]
@jojochuang please let me know if you are OK with me updating the PR with the minor refactoring I suggested

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java [hunk 3]
```diff
@@ -170,6 +181,21 @@ public void write(DataOutput out) throws IOException {
...
+        .setOwnerId(getOwnerId())
+        .setSecretKeyId(ProtobufUtils.toProtobuf(getSecretKeyId()))
+        .setExpiryDate(getExpiryDate())
+        .setMaxLength(maxLength);
+    // Add access mode allowed
+    for (AccessModeProto mode : modes) {
+      builder.addModes(AccessModeProto.valueOf(mode.name()));
+    }
+    return builder.build().toByteArray();
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java:197]: `getBytes()` is overridden (optimized) only for block token.  Did you intend to do the same for container token?

**Review by adoroszlai — COMMENTED** [review]
Thanks @jojochuang for updating the patch.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java (hunk 3 above)
Comment (jojochuang) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/OzoneBlockTokenIdentifier.java]: Good idea.

Also refactored OzoneBlockTokenIdentifier write() to use getBytes().

**Review by adoroszlai — APPROVED** [review]
Thanks @jojochuang for improving the patch, LGTM.

**Pull request merged**